source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc

```diff
@@ -275,37 +263,48 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
   // It's possible that the cluster specified in the filter configuration no longer exists due to a
   // CDS removal.
   if (cm_.get(cluster) == nullptr) {
-    // TODO(dio): Add stats and tracing related to this.
+    // TODO(dio): Add stats related to this.
     ENVOY_LOG(debug, "ext_authz cluster '{}' does not exist", cluster);
     callbacks_->onComplete(std::make_unique<Response>(errorResponse()));
-    span_->setTag(Tracing::Tags::get().Error, Tracing::Tags::get().True);
-    span_->finishSpan();
     callbacks_ = nullptr;
-    span_ = nullptr;
   } else {
-    span_->injectContext(message->headers());
-    request_ = cm_.httpAsyncClientForCluster(cluster).send(
-        std::move(message), *this,
-        Http::AsyncClient::RequestOptions().setTimeout(config_->timeout()));
+    auto options = Http::AsyncClient::RequestOptions()
+                       .setTimeout(config_->timeout())
+                       .setParentSpan(parent_span)
+                       .setChildSpanName(config_->tracingName());
+
+    request_ = cm_.httpAsyncClientForCluster(cluster).send(std::move(message), *this, options);
   }
 }
 
 void RawHttpClientImpl::onSuccess(const Http::AsyncClient::Request&,
                                   Http::ResponseMessagePtr&& message) {
   callbacks_->onComplete(toResponse(std::move(message)));
-  span_->finishSpan();
   callbacks_ = nullptr;
-  span_ = nullptr;
 }
 
 void RawHttpClientImpl::onFailure(const Http::AsyncClient::Request&,
                                   Http::AsyncClient::FailureReason reason) {
   ASSERT(reason == Http::AsyncClient::FailureReason::Reset);
   callbacks_->onComplete(std::make_unique<Response>(errorResponse()));
-  span_->setTag(Tracing::Tags::get().Error, Tracing::Tags::get().True);
-  span_->finishSpan();
   callbacks_ = nullptr;
-  span_ = nullptr;
+}
+
+void RawHttpClientImpl::onBeforeFinalizeUpstreamSpan(
+    Tracing::Span& span, const Http::ResponseHeaderMap* response_headers, bool success) {
+  if (success && response_headers != nullptr) {
+    uint64_t status_code{};
+    if (!absl::SimpleAtoi(response_headers->getStatusValue(), &status_code)) {
+      span.setTag(Tracing::Tags::get().Error, Tracing::Tags::get().True);
```

Comment:
The base async_client impl sets error on reset so I don't think this is needed? And per other comment if we simplify that response_headers != null means that request was not reset, I think this all gets a lot simpler?

Reply:
Yes. Thanks for this.